crates/store/src/genesis/config/mod.rs

```diff
@@ -0,0 +1,327 @@
+                .account_type(account_type)
+                .storage_mode(account_storage_mode)
+                .with_auth_component(auth)
+                .with_component(component)
+                .build()?;
+
+            faucet_account.apply_delta(&AccountDelta::new(
+                faucet_account.id(),
+                AccountStorageDelta::new(),
+                AccountVaultDelta::default(),
+                Felt::ONE,
+            )?)?;
+
+            if faucets.insert(symbol.clone(), faucet_account.id()).is_some() {
+                return Err(GenesisConfigError::DuplicateFaucetDefinition { symbol: token_symbol });
+            }
+
+            secrets.push((
+                format!("faucet_{symbol}.mac", symbol = symbol.to_lowercase()),
+                faucet_account.clone(),
+                secret_key,
+                faucet_account_seed,
+            ));
+
+            all_accounts.push(faucet_account);
+        }
+
+        let zero_padding_width = usize::ilog10(std::cmp::max(10, wallet_configs.len())) as usize;
+
+        // then setup all wallet accounts, which reference the faucet's for their provided assets
+        for (index, WalletConfig { has_updatable_code, storage_mode, assets }) in
+            wallet_configs.into_iter().enumerate()
+        {
+            let mut rng = ChaCha20Rng::from_seed(rand::random());
+            let secret_key = SecretKey::with_rng(&mut get_rpo_random_coin(&mut rng));
+            let auth = AuthScheme::RpoFalcon512 { pub_key: secret_key.public_key() };
+            let init_seed: [u8; 32] = rng.random();
+
+            let account_type = if has_updatable_code {
+                AccountType::RegularAccountUpdatableCode
+            } else {
+                AccountType::RegularAccountImmutableCode
+            };
+            let account_storage_mode = storage_mode.into();
+            let (mut wallet_account, wallet_account_seed) =
+                create_basic_wallet(init_seed, auth, account_type, account_storage_mode)?;
+
+            // Add fungible assets.
+            let fungible_assets = prepare_asset_update(assets, &faucets)?;
+
+            // Force the account nonce to 1.
+            //
+            // By convention, a nonce of zero indicates a freshly generated local account that has
+            // yet to be deployed. An account is deployed onchain along with its first
+            // transaction which results in a non-zero nonce onchain.
+            //
+            // The genesis block is special in that accounts are "deplyed" without transactions and
+            // therefore we need bump the nonce manually to uphold this invariant.
+            let delta = AccountDelta::new(
+                wallet_account.id(),
+                AccountStorageDelta::default(),
+                AccountVaultDelta::new(fungible_assets, NonFungibleAssetDelta::default()),
+                Felt::ONE,
+            )?;
+
+            wallet_account.apply_delta(&delta)?;
```

Comment:
Something that just occurred to me.. Do we ever subtract these assets from the original faucets?

I don't think we do currently. I imagine this sort of book-keeping is usually enforced by virtue of the note's system. But iiuc we would break the stated max-supply invariant because the faucet would be at max supply but wallets would have additional amounts?

Reply:
@bobbinth maybe you have some ideas here? Or is it just a literal case of subtracting the amounts from the faucets already? Which probably means we have to re-order some of this.